cpp/src/parquet/stream_reader.cc

```diff
@@ -0,0 +1,224 @@
+  std::int32_t tmp;
+  read_<Int32Reader>(tmp);
+  v = static_cast<std::int16_t>(tmp);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::uint16_t& v) {
+  checkColumn_(Type::INT32, ConvertedType::UINT_16);
+  std::int32_t tmp;
+  read_<Int32Reader>(tmp);
+  v = static_cast<std::uint16_t>(tmp);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::int32_t& v) {
+  checkColumn_(Type::INT32, ConvertedType::INT_32);
+  read_<Int32Reader>(v);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::uint32_t& v) {
+  checkColumn_(Type::INT32, ConvertedType::UINT_32);
+  read_<Int32Reader>(reinterpret_cast<std::int32_t&>(v));
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::int64_t& v) {
+  checkColumn_(Type::INT64, ConvertedType::INT_64);
+  read_<Int64Reader>(v);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::uint64_t& v) {
+  checkColumn_(Type::INT64, ConvertedType::UINT_64);
+  read_<Int64Reader>(reinterpret_cast<std::int64_t&>(v));
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::chrono::milliseconds& v) {
+  checkColumn_(Type::INT64, ConvertedType::TIMESTAMP_MILLIS);
+  std::int64_t tmp;
+  read_<Int64Reader>(tmp);
+  v = std::chrono::milliseconds{tmp};
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::chrono::microseconds& v) {
+  checkColumn_(Type::INT64, ConvertedType::TIMESTAMP_MICROS);
+  std::int64_t tmp;
+  read_<Int64Reader>(tmp);
+  v = std::chrono::microseconds{tmp};
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(float& v) {
+  checkColumn_(Type::FLOAT, ConvertedType::NONE);
+  read_<FloatReader>(v);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(double& v) {
+  checkColumn_(Type::DOUBLE, ConvertedType::NONE);
+  read_<DoubleReader>(v);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(char& v) {
+  checkColumn_(Type::FIXED_LEN_BYTE_ARRAY, ConvertedType::NONE, 1);
+  FixedLenByteArray flba;
+  read_(flba);
+  v = static_cast<char>(flba.ptr[0]);
+  return *this;
+}
+
+StreamReader& StreamReader::operator>>(std::string& v) {
+  checkColumn_(Type::BYTE_ARRAY, ConvertedType::UTF8);
+  ByteArray ba;
+  read_(ba);
+  v = std::string(reinterpret_cast<const char*>(ba.ptr), ba.len);
+  return *this;
+}
+
+void StreamReader::readFixedLength_(char* ptr, const int len) {
+  checkColumn_(Type::FIXED_LEN_BYTE_ARRAY, ConvertedType::NONE, len);
+  FixedLenByteArray flba;
+  read_(flba);
+  std::memcpy(ptr, flba.ptr, len);
+}
+
+void StreamReader::read_(ByteArray& v) {
+  const auto& node{nodes_[column_index_]};
+  auto* const reader{
+      static_cast<ByteArrayReader*>(column_readers_[column_index_++].get())};
+  std::int64_t values_read;
+
+  reader->ReadBatch(1, nullptr, nullptr, &v, &values_read);
+
+  if (values_read != 1) {
+    throw ParquetException("Failed to read value for column '" + node->name() + "'");
+  }
+}
+
+void StreamReader::read_(FixedLenByteArray& v) {
```

Comment:
use pointer for output parameters when it isn't required for the API https://google.github.io/styleguide/cppguide.html#Output_Parameters